plugins/flutter_plugins/path_provider_fde/linux/path_provider_plugin.cc

```diff
@@ -0,0 +1,161 @@
+                << std::endl;
+    }
+  }
+}
+
+class PathProviderPlugin : public flutter::Plugin {
+ public:
+  static void RegisterWithRegistrar(flutter::PluginRegistrar *registrar);
+
+  virtual ~PathProviderPlugin();
+
+ private:
+  PathProviderPlugin();
+  std::string exe_name;
+  std::string homedir_name;
+  // Called when a method is called on plugin channel;
+  void HandleMethodCall(
+      const flutter::MethodCall<EncodableValue> &method_call,
+      std::unique_ptr<flutter::MethodResult<EncodableValue>> result);
+};
+
+// static
+void PathProviderPlugin::RegisterWithRegistrar(
+    flutter::PluginRegistrar *registrar) {
+  auto channel = std::make_unique<flutter::MethodChannel<EncodableValue>>(
+      registrar->messenger(), "plugins.flutter.io/path_provider",
+      &flutter::StandardMethodCodec::GetInstance());
+
+  // Uses new instead of make_unique due to private constructor.
+  std::unique_ptr<PathProviderPlugin> plugin(new PathProviderPlugin());
+
+  channel->SetMethodCallHandler(
+      [plugin_pointer = plugin.get()](const auto &call, auto result) {
+        plugin_pointer->HandleMethodCall(call, std::move(result));
+      });
+
+  registrar->AddPlugin(std::move(plugin));
+}
+
+PathProviderPlugin::PathProviderPlugin() {
+  exe_name = GetExecutableName();
+  const char *homedir;
+  homedir = getenv("HOME");
+  homedir_name = std::string(homedir);
+}
+
+PathProviderPlugin::~PathProviderPlugin() = default;
+
+void PathProviderPlugin::HandleMethodCall(
+    const flutter::MethodCall<EncodableValue> &method_call,
+    std::unique_ptr<flutter::MethodResult<EncodableValue>> result) {
+  if (method_call.method_name().compare("getTemporaryDirectory") == 0) {
+    std::string result_path = "/tmp";
+    flutter::EncodableValue response(result_path);
+    result->Success(&response);
+  } else if (method_call.method_name().compare(
+                 "getApplicationSupportDirectory") == 0) {
+    if (homedir_name.empty()) {
+      result->Error("Unable to get usr home directory");
+      return;
+    }
+    // Use the executable name as the subdirectory for now.
+    if (exe_name.empty()) {
+      result->Error("Unable to get exe name");
+      return;
+    }
+    std::ostringstream response_stream;
+    response_stream << homedir_name << "/.cache/" << exe_name;
+    EnsureCreated(response_stream.str());
+    flutter::EncodableValue response(response_stream.str());
+    result->Success(&response);
+  } else if (method_call.method_name().compare(
+                 "getApplicationDocumentsDirectory") == 0) {
+    if (homedir_name.empty()) {
+      result->Error("Unable to get usr home directory");
+      return;
+    }
+    // Use the executable name as the subdirectory for now.
+    if (exe_name.empty()) {
+      result->Error("Unable to get exe name");
+      return;
+    }
+    std::ostringstream response_stream;
+    response_stream << homedir_name << "/.local/share/" << exe_name;
```

Comment:
The primary distinction between `getApplicationDocumentsDirectory` and the other application data directories in `path_provider` is that this one should be user visible, so a dot directory doesn't seem like the right semantics. The way I've interpreted this on the other desktop platform is to use the general user documents directory. It looks like in XDG terms that would be XDG_DOCUMENTS_DIR.

Also, I haven't been making application-specific subdirs for this one on other desktop platforms. What the right behavior here is isn't obvious, since the API was designed for mobile where everything is sandboxed at the application level in most cases anyway, but for now the approach I've been taking is to just return the documents directory as-is. I'd like to be consistent, and if there's feedback that that's unexpected, change all of the desktop platforms together.